### PR TITLE
Support for using custom external diff command

### PR DESCRIPTION
**Pull request by wrobelda**
This allows for using more robust,  content-aware comparison tools, e.g. for XML, JSON, etc.

### CODE REVIEW

File: urlwatch [hunk 1]
```diff
@@ -370,6 +394,19 @@ if __name__ == '__main__':
             if display_errors:
                 details += foutput('error', job, (repr(error) +
                         '\n' + str(error)).strip(), summary)
+        except OSError, error:
+            log.error('got OSError while trying to execute diff command: %s' % error)
```
Comment (thp) [inline urlwatch]: The `log.error()` function does `%`-formatting by itself, so `log.error('got ... %s', error)` would be better here (and below).

File: urlwatch [hunk 2]
```diff
@@ -370,6 +394,19 @@ if __name__ == '__main__':
...
+            if display_errors:
+                details += foutput('error', job, error, summary)
+        except KeyError, error:
+            log.error('got error while attempting to replace diff command\'s placeholder(s): %s' % error)
+            if display_errors:
+                details += foutput('error', job, error, summary)
+        except subprocess.CalledProcessError, error:
+            log.error('got error while executing diff command %s: %s' % (error.cmd, error.output))
```
Comment (thp) [inline urlwatch]: Here it would be `log.error('got error while executing diff command %s: %s', error.cmd, error.output)`

File: README.md [hunk 3]
```diff
@@ -62,6 +62,12 @@ A: Indeed there is. See the example hooks.py file.
 Q: What about badly-formed HTML (long lines, etc..)?
 A: Use python-utidylib. See the example hooks.py file.
 
+Q: Is there a way to compare the data using a different diff tool (eg. syntax-aware XML diff)?
+A: Use "--diff-command=COMMAND", where COMMAND is a full filesystem path to the
+   tool followed by its arguments. Substitute compared files' names with '{OLD_DATA}'
+   and '{NEW_DATA}' tokens. This doesn't support pipelining of multiple
+   commands - put these in a sh script and use this instead.
```
Comment (thp) [inline README.md]: Given that we don't support pipelining (should we? subprocess has `shell=True` for that), maybe we shouldn't request `{OLD_DATA}` and `{NEW_DATA}` but just pass in the first and second argument? (and if the diff utility needs different parameter ordering and/or additional commands, just let a shell script deal with those, and call the shell script with the two positional arguments)

**Comment by thp** [comment]
Added some comments. Looks like a really useful feature, wonder if overriding the diff utility should be done per-URL (e.g. I would want to use one diff utility for a XML document, and a different one for something else, e.g. word diff).

**Comment by thp** [comment]
@wrobelda: Any chance to update this PR with the comments from above?

**Comment by wrobelda** [comment]
Sure I will, was too busy last weeks. Will do this weekend.

**Comment by thp** [comment]
Please update this so it fits the new YAML style model (maybe have a per-URL configuration option that can specify the external diff command)?

**Comment by thp** [comment]
As mentioned in #21, maybe this can be implemented as new Reporter subclass for urlwatch 2?

**Comment by wrobelda** [comment]
Sorry for late reply - I have been out for holidays starting mid-December. I will look into this over this weekend or following week.

**Comment by thp** [comment]
Closing this for now, if you have time to update this and rebase against the current version, feel free to reopen this one or open a new pull request.

**Pull request closed without merging**